In [3]:
import numpy as np

def print_matrix(matrix, label):
    print(label)
    for row in matrix:
        print(row)

def cholesky_decomposition(A):
    n = len(A)
    L = np.zeros((n, n))

    for i in range(n):
        for j in range(i + 1):
            if i == j:
                L[i, j] = np.sqrt(A[i, i] - np.sum(L[i, :j]**2))
            else:
                L[i, j] = (A[i, j] - np.sum(L[i, :j] * L[j, :j])) / L[j, j]

    return L

def verify_cholesky_decomposition(A, L):
    reconstructed_A = np.dot(L, L.T)
    return np.allclose(A, reconstructed_A)

# Example with random input
# np.random.seed(42)  # Set seed for reproducibility
n = 3
A = np.random.randint(1, 10, size=(n, n))
A = A @ A.T  # Make a symmetric positive definite matrix
print_matrix(A, "Original Matrix A:")

# Perform Cholesky decomposition
L = cholesky_decomposition(A)

print_matrix(L, "Lower Triangular Matrix L:")

# Verify A = LL^T
verification_result = verify_cholesky_decomposition(A, L)
print("Verification Result (A = LL^T):", verification_result)


Original Matrix A:
[ 6 16 11]
[ 16 107  32]
[11 32 26]
Lower Triangular Matrix L:
[2.44948974 0.         0.        ]
[6.53197265 8.02080628 0.        ]
[4.4907312  0.33246865 2.39223701]
Verification Result (A = LL^T): True


<p> Vishal </p> 

Original Matrix A:
[ 6 16 11]
[ 16 107  32]
[11 32 26]
Lower Triangular Matrix L:
[2.44948974 0.         0.        ]
[6.53197265 8.02080628 0.        ]
[4.4907312  0.33246865 2.39223701]
Verification Result (A = LL^T): True

<p> Anshuman </p>
Original Matrix A:
[75 67 52]
[67 65 63]
[52 63 90]
Lower Triangular Matrix L:
[8.66025404 0.         0.        ]
[7.73649361 2.2686266  0.        ]
[6.0044428  7.2936933  0.86527722]
Verification Result (A = LL^T): True

<p> Amit Bhai </p>

Original Matrix A:
[66 94 73]
[ 94 149 125]
[ 73 125 110]
<br>
Lower Triangular Matrix L:
[8.1240384 0.        0.       ]
[11.57060015  3.88860028  0.        ]
[8.98567884 5.40819357 0.09496334]
<br>
Verification Result (A = LL^T): True